File: client/flutter/div_expressions_resolver/android/src/main/kotlin/tech/divkit/flutter/div_expressions_resolver/Protocol.kt
import android.util.Log
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MessageCodec
import io.flutter.plugin.common.StandardMessageCodec
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

private fun wrapResult(result: Any?): List<Any?> {
  return listOf(result)
}

private fun wrapError(exception: Throwable): List<Any?> {
  return if (exception is FlutterError) {
    listOf(
      exception.code,
      exception.message,
      exception.details
    )
  } else {
    listOf(
      exception.javaClass.simpleName,
      exception.toString(),
      "Cause: " + exception.cause + ", Stacktrace: " + Log.getStackTraceString(exception)
    )
  }
}

/**
 * Error class for passing custom error details to Flutter via a thrown PlatformException.
 * @property code The error code.
 * @property message The error message.
 * @property details The error details. Must be a datatype supported by the api codec.
 */
class FlutterError (
  val code: String,
  override val message: String? = null,
  val details: Any? = null
) : Throwable()

data class IntColor (
  val value: Int
) {
  companion object {
    @Suppress("UNCHECKED_CAST")
    fun fromList(list: List<Any?>): IntColor {
      val value = list[0].let { if (it is Long) it.toInt() else it as Int }
      return IntColor(value)
    }
  }
  fun toList(): List<Any?> {
    return listOf<Any?>(
      value,
    )
  }
}

data class StringUrl (
  val value: String
) {
  companion object {
    @Suppress("UNCHECKED_CAST")
    fun fromList(list: List<Any?>): StringUrl {
      val value = list[0] as String
      return StringUrl(value)
    }
  }
  fun toList(): List<Any?> {
    return listOf<Any?>(
      value,
    )
  }
}

@Suppress("UNCHECKED_CAST")
private object NativeDivExpressionsResolverCodec : StandardMessageCodec() {
  override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? {
    return when (type) {
      128.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          IntColor.fromList(it)
        }
      }
      129.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          StringUrl.fromList(it)
        }
      }
      else -> super.readValueOfType(type, buffer)
    }
  }
  override fun writeValue(stream: ByteArrayOutputStream, value: Any?)   {
    when (value) {
      is IntColor -> {
        stream.write(128)
        writeValue(stream, value.toList())
      }
      is StringUrl -> {
        stream.write(129)
        writeValue(stream, value.toList())
      }
      else -> super.writeValue(stream, value)
    }
  }
}

interface NativeDivExpressionsResolver {
  fun resolve(expression: String, context: Map<String, Any>, callback: (Result<Any?>) -> Unit)
  fun clearVariables(callback: (Result<Unit>) -> Unit)

  companion object {
    /** The codec used by NativeDivExpressionsResolver. */
    val codec: MessageCodec<Any?> by lazy {
      NativeDivExpressionsResolverCodec
    }
    /** Sets up an instance of `NativeDivExpressionsResolver` to handle messages through the `binaryMessenger`. */
    @Suppress("UNCHECKED_CAST")
    fun setUp(binaryMessenger: BinaryMessenger, api: NativeDivExpressionsResolver?) {
      run {
        val channel = BasicMessageChannel<Any?>(binaryMessenger, "div_expressions_resolver.NativeDivExpressionsResolver.resolve", codec)
        if (api != null) {
          channel.setMessageHandler { message, reply ->
            val args = message as List<Any?>
            val expressionArg = args[0] as String
            val contextArg = args[1] as Map<String, Any>
            api.resolve(expressionArg, contextArg) { result: Result<Any?> ->
              val error = result.exceptionOrNull()
              if (error != null) {
                reply.reply(wrapError(error))
              } else {
                val data = result.getOrNull()
                reply.reply(wrapResult(data))
              }
            }
          }
        } else {
          channel.setMessageHandler(null)
        }
      }
      run {
        val channel = BasicMessageChannel<Any?>(binaryMessenger, "div_expressions_resolver.NativeDivExpressionsResolver.clearVariables", codec)
        if (api != null) {
          channel.setMessageHandler { _, reply ->
            api.clearVariables() { result: Result<Unit> ->
              val error = result.exceptionOrNull()
              if (error != null) {
                reply.reply(wrapError(error))
              } else {
                reply.reply(wrapResult(null))
              }
            }
          }
        } else {
          channel.setMessageHandler(null)
        }
      }
    }
  }
}

File: client/flutter/div_expressions_resolver/ios/Classes/Protocol.swift
import Foundation
#if os(iOS)
import Flutter
#elseif os(macOS)
import FlutterMacOS
#else
#error("Unsupported platform.")
#endif

private func wrapResult(_ result: Any?) -> [Any?] {
  return [result]
}

private func wrapError(_ error: Any) -> [Any?] {
  if let flutterError = error as? FlutterError {
    return [
      flutterError.code,
      flutterError.message,
      flutterError.details
    ]
  }
  return [
    "\(error)",
    "\(type(of: error))",
    "Stacktrace: \(Thread.callStackSymbols)"
  ]
}

private func isNullish(_ value: Any?) -> Bool {
  return value is NSNull || value == nil
}

private func nilOrValue<T>(_ value: Any?) -> T? {
  if value is NSNull { return nil }
  return value as! T?
}

extension FlutterError: Error {}

struct IntColor {
  var value: Int

  static func fromList(_ list: [Any?]) -> IntColor? {
    let value = list[0] is Int ? list[0] as! Int : Int(list[0] as! Int64)

    return IntColor(
      value: value
    )
  }
  func toList() -> [Any?] {
    return [
      value,
    ]
  }
}

struct StringUrl {
  var value: String

  static func fromList(_ list: [Any?]) -> StringUrl? {
    let value = list[0] as! String

    return StringUrl(
      value: value
    )
  }
  func toList() -> [Any?] {
    return [
      value,
    ]
  }
}

private class NativeDivExpressionsResolverCodecReader: FlutterStandardReader {
  override func readValue(ofType type: UInt8) -> Any? {
    switch type {
      case 128:
        return IntColor.fromList(self.readValue() as! [Any?])
      case 129:
        return StringUrl.fromList(self.readValue() as! [Any?])
      default:
        return super.readValue(ofType: type)
    }
  }
}

private class NativeDivExpressionsResolverCodecWriter: FlutterStandardWriter {
  override func writeValue(_ value: Any) {
    if let value = value as? IntColor {
      super.writeByte(128)
      super.writeValue(value.toList())
    } else if let value = value as? StringUrl {
      super.writeByte(129)
      super.writeValue(value.toList())
    } else {
      super.writeValue(value)
    }
  }
}

private class NativeDivExpressionsResolverCodecReaderWriter: FlutterStandardReaderWriter {
  override func reader(with data: Data) -> FlutterStandardReader {
    return NativeDivExpressionsResolverCodecReader(data: data)
  }

  override func writer(with data: NSMutableData) -> FlutterStandardWriter {
    return NativeDivExpressionsResolverCodecWriter(data: data)
  }
}

class NativeDivExpressionsResolverCodec: FlutterStandardMessageCodec {
  static let shared = NativeDivExpressionsResolverCodec(readerWriter: NativeDivExpressionsResolverCodecReaderWriter())
}

protocol NativeDivExpressionsResolver {
  func resolve(expression: String, context: [String: Any], completion: @escaping (Result<Any?, Error>) -> Void)
  func clearVariables(completion: @escaping (Result<Void, Error>) -> Void)
}

public class NativeDivExpressionsResolverSetup {
  /// The codec used by NativeDivExpressionsResolver.
  static var codec: FlutterStandardMessageCodec { NativeDivExpressionsResolverCodec.shared }
  /// Sets up an instance of `NativeDivExpressionsResolver` to handle messages through the `binaryMessenger`.
  static func setUp(binaryMessenger: FlutterBinaryMessenger, api: NativeDivExpressionsResolver?) {
    let resolveChannel = FlutterBasicMessageChannel(name: "div_expressions_resolver.NativeDivExpressionsResolver.resolve", binaryMessenger: binaryMessenger, codec: codec)
    if let api = api {
      resolveChannel.setMessageHandler { message, reply in
        let args = message as! [Any?]
        let expressionArg = args[0] as! String
        let contextArg = args[1] as! [String: Any]
        api.resolve(expression: expressionArg, context: contextArg) { result in
          switch result {
            case .success(let res):
              reply(wrapResult(res))
            case .failure(let error):
              reply(wrapError(error))
          }
        }
      }
    } else {
      resolveChannel.setMessageHandler(nil)
    }
    let clearVariablesChannel = FlutterBasicMessageChannel(name: "div_expressions_resolver.NativeDivExpressionsResolver.clearVariables", binaryMessenger: binaryMessenger, codec: codec)
    if let api = api {
      clearVariablesChannel.setMessageHandler { _, reply in
        api.clearVariables() { result in
          switch result {
            case .success:
              reply(wrapResult(nil))
            case .failure(let error):
              reply(wrapError(error))
          }
        }
      }
    } else {
      clearVariablesChannel.setMessageHandler(nil)
    }
  }
}
